### PR TITLE
Use `TheoryData` wherever possible

### PR DESCRIPTION
**Pull request by bastbu**
# PR for issue #1018

## What is being addressed

We use `TheoryData<>` wherever possible instead of more weakly typed `IEnumerable<object[]>` for theory data to get stronger correctness guarantees at compile time.

### CODE REVIEW

File: Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs [hunk 1]
```diff
@@ -95,30 +96,38 @@ public void When_Data_Message_Encountered_Should_Find_Duplicates_For_Different_D
...
+            (new DevEui(0), new Mic(0), 0, "1"), // a non-relevant field should not influence the key
+            (new DevEui(0x1010101010101010UL), new Mic(0), 0, null),
+            (new DevEui(0), new Mic(1), 0, null),
+            (new DevEui(0), new Mic(0), 1, null)
+        };
+
+        public static TheoryData<DataMessageKeyHolder, DevEui, Mic, ushort, string?> CreateKeyDataMessagesTheoryData =>
+            TheoryDataFactory.From(from dataPoint in RawCreateKeyDataMessagesTheoryData
+                                   select (new DataMessageKeyHolder(new ConcentratorDeduplication.DataMessageKey(dataPoint.DevEui, dataPoint.Mic, dataPoint.FCnt)),
+                                           dataPoint.DevEui, dataPoint.Mic, dataPoint.FCnt, dataPoint.FieldNotUsedInKey));
```
Comment (spygi) [inline Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs:117]: Not sure if all these indirections, additional type etc make the test easier to understand..why not just use the strong types for DevEui and Mic?

```suggestion
               public static TheoryData CreateKeyDataMessagesTheoryData
                    => TheoryDataFactory.From<ConcentratorDeduplication.DataMessageKey, DevEui, Mic, ushort, string?>(
                        new (ConcentratorDeduplication.DataMessageKey, DevEui, Mic, ushort, string?)[]
                        {
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), new DevEui(0), new Mic(0), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), new DevEui(0), new Mic(0), 0, "1"), // a non-relevant field should not influence the key
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0x1010101010101010UL), new Mic(0), 0), new DevEui(0x1010101010101010UL), new Mic(0), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(1), 0), new DevEui(0), new Mic(1), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 1), new DevEui(0), new Mic(0), 1, null)
                        }
                    );
```

**Review by spygi — APPROVED** [review]
LGTM - thank you for this change!

File: Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs (hunk 1 above)
Comment (bastbu) [reply Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs:118]: Added the clarification in the code with 67d7fa00cad9d5f9a86dec47326e92951cfb4feb

Comment (spygi) [reply Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs:118]: ok, I see your point about the return type now (though I would say that we get already some type guarantees from the `TheoryDataFactory.From` method). Thanks for the explanation :)

**Pull request merged**